File: editor/ARCed/src/Core/ARC_Data.py
from struct import pack, unpack
import types

class ARC_Data(object):

    _VERSION = "\x01\x00" # 1.0
    
    _TYPES = {
        "NoneClass" : chr(0x10),
        "FalseClass" : chr(0x11),
        "TrueClass" : chr(0x12),
        "Fixnum" : chr(0x21),
        "Bignum" : chr(0x22),
        "Float" : chr(0x23),
        "String" : chr(0x30),
        "Array" : chr(0x40),
        "Hash" : chr(0x41),
        "Object" : chr(0x00)
        }
    
    _io = None
    _strings = [None]
    _arrays = [None]
    _hashes = [None]
    _objects = [None]
    _class_path_redirects = {}
    _extended_namespace = {}
    
    @staticmethod
    def dump(io, obj, redirects={}):
        ARC_Data._class_path_redirects = redirects
        ARC_Data._io = io
        ARC_Data._io.write(ARC_Data._VERSION)
        try:
            ARC_Data._dump(obj)
        except Exception:
            try:
                print "stream position: %s" % ARC_Data._io.tell()
            except Exception:
                pass
            raise
        ARC_Data._reset()
    
    
    @staticmethod
    def load(io, redirects={}, extended_namespace={}):
        ARC_Data._class_path_redirects = redirects
        ARC_Data._extended_namespace = dict(extended_namespace)
        ARC_Data._extended_namespace.update(globals())
        extended_namespace = {}
        ARC_Data._io = io
        version = ARC_Data._io.read(2)
        if ARC_Data._VERSION != version:
            raise TypeError("Error: version mismatch! Expected: %s Found: %s" %(repr(ARC_Data._VERSION), repr(version)))
        try:
            data = ARC_Data._load()
        except Exception:
            try:
                print "stream position: %s" % ARC_Data._io.tell()
            except Exception:
                pass
            raise
        ARC_Data._reset()
        return data
    
    
    @staticmethod
    def _reset():
        ARC_Data._io = None
        ARC_Data._strings = [None]
        ARC_Data._arrays = [None]
        ARC_Data._hashes = [None]
        ARC_Data._objects = [None]
        ARC_Data._class_path_redirects = {}
        ARC_Data._extended_namespace = {}
    

    @staticmethod
    def __get_class_path(name):
        if ARC_Data._class_path_redirects.has_key(name):
            return ARC_Data._class_path_redirects[name]
        else: 
            return name
    
    
    @staticmethod
    def __get_class_object(class_path):
        classes = class_path.split("::")
        if not ARC_Data._extended_namespace.has_key(classes[0]):
            raise TypeError("Class not defined: %s" % classes[0])    
        classe = ARC_Data._extended_namespace[classes.pop(0)]
        for c in classes:
            if not classe.__dict__.has_key(c):
                raise TypeError("Class not defined: %s" % c)
            classe = classe.const_get(c.to_sym)
        return classe
    

    @staticmethod
    def __try_map(data, obj):
        if obj in data:
            index = data.index(obj)
        else:
            index = None
        if index == None:
            ARC_Data.__dump_int32(len(data))
            data.append(obj)
            return True
        ARC_Data.__dump_int32(index)
        return False
    

    @staticmethod
    def __map(data, obj):
        data.append(obj)
    

    @staticmethod
    def __find_mapped(data, id_num):
        if id_num < len(data):
            return data[id_num]
        else:
            return None
    
    
    @staticmethod
    def __dump_int32(obj):
        ARC_Data._io.write(pack("<I", obj))
    

    @staticmethod
    def __load_int32():
        return unpack("<I", ARC_Data._io.read(4))[0]
    
    
    @staticmethod
    def _dump(obj):
        if obj == None:
            return ARC_Data._dump_none(obj)
        if obj == False:
            return ARC_Data._dump_false(obj) 
        if obj == True:
            return ARC_Data._dump_true(obj)
        if isinstance(obj, types.IntType):
            return ARC_Data._dump_fixnum(obj)
        if isinstance(obj, types.LongType):
            return ARC_Data._dump_bignum(obj)
        if isinstance(obj, types.FloatType):
            return ARC_Data._dump_float(obj)
        if isinstance(obj, types.StringTypes):
            return ARC_Data._dump_string(obj)
        if isinstance(obj, (types.ListType, types.TupleType)):
            return ARC_Data._dump_array(obj)
        if isinstance(obj, types.DictType):
            return ARC_Data._dump_hash(obj)
        if isinstance(obj, object):
            return ARC_Data._dump_object(obj)
        raise TypeError("Error: %s cannot be dumped!" % obj.__class__)
    
    @staticmethod
    def _load():
        type_id = ARC_Data._io.read(1)
        if type_id == ARC_Data._TYPES["NoneClass"]:
            return ARC_Data._load_none()
        elif type_id == ARC_Data._TYPES["FalseClass"]:
            return ARC_Data._load_false()
        elif type_id == ARC_Data._TYPES["TrueClass"]:
            return ARC_Data._load_true()
        elif type_id == ARC_Data._TYPES["Fixnum"]:
            return ARC_Data._load_fixnum()
        elif type_id == ARC_Data._TYPES["Bignum"]:
            return ARC_Data._load_bignum()
        elif type_id == ARC_Data._TYPES["Float"]:
            return ARC_Data._load_float()
        elif type_id == ARC_Data._TYPES["String"]:
            return ARC_Data._load_string()
        elif type_id == ARC_Data._TYPES["Array"]:
            return ARC_Data._load_array()
        elif type_id == ARC_Data._TYPES["Hash"]:
            return ARC_Data._load_hash()
        elif type_id == ARC_Data._TYPES["Object"]:
            return ARC_Data._load_object()
        
        raise TypeError("Error: Unknown type 0x%02X detected!" % ord(type_id))
    
    
    @staticmethod
    def _dump_none(obj):
        ARC_Data._io.write(ARC_Data._TYPES["NoneClass"])
    

    @staticmethod
    def _dump_false(obj):
        ARC_Data._io.write(ARC_Data._TYPES["FalseClass"])
    

    @staticmethod
    def _dump_true(obj):
        ARC_Data._io.write(ARC_Data._TYPES["TrueClass"])
    
    
    @staticmethod
    def _dump_fixnum(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Fixnum"])
        ARC_Data.__dump_int32(obj)
    
    
    @staticmethod
    def _dump_bignum(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Bignum"])
        ARC_Data.__dump_int32(obj) # our C++ implementation uses a "long" of 32 bit
    
    
    @staticmethod
    def _dump_float(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Float"])
        ARC_Data._io.write(pack("<f", obj))
    
    
    @staticmethod
    def _dump_string(obj):
        ARC_Data._io.write(ARC_Data._TYPES["String"])
        if len(obj) > 0:
            if not ARC_Data.__try_map(ARC_Data._strings, obj): # abort if object has already been mapped
                return
            ARC_Data.__dump_int32(len(obj))
            ARC_Data._io.write(obj)
        else:
            ARC_Data.__dump_int32(0)
        
    
    
    @staticmethod
    def _dump_array(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Array"])
        if len(obj) > 0:
            if not ARC_Data.__try_map(ARC_Data._arrays, obj): # abort if object has already been mapped
                return 
            ARC_Data.__dump_int32(len(obj))
            for value in obj:
                ARC_Data._dump(value)
        else:
            ARC_Data.__dump_int32(0)
        
    
    
    @staticmethod
    def _dump_hash(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Hash"])
        if len(obj) > 0:
            if not ARC_Data.__try_map(ARC_Data._hashes, obj): # abort if object has already been mapped
                return 
            ARC_Data.__dump_int32(len(obj))
            for key, value in obj.items():
                ARC_Data._dump(key)
                ARC_Data._dump(value)
        else:
            ARC_Data.__dump_int32(0)
        
    
    @staticmethod
    def _request_instance_variables(obj):
        if hasattr(obj, "_arc_instance_variables"):
            return obj._arc_instance_variables
        else:
            l = []
            for key, value in obj.__dict__.items():
                if key[0] != "_":
                    if not isinstance(value, (types.FunctionType, types.ClassType, types.MethodType, 
                                              types.ModuleType, types.SliceType, types.LambdaType, 
                                              types.GeneratorType)):
                        l.append(key)
            return l
    
    @staticmethod
    def _dump_object(obj):
        ARC_Data._io.write(ARC_Data._TYPES["Object"])
        if hasattr(obj, "_arc_class_path"):
            klass_path = obj._arc_class_path
        else:
            mod_names = obj.__class__.__module__.split(".")
            mod_name =  mod_names[0]
            if len(mod_names) > 1:
                for name in mod_names[1:]:
                    mod_name += "::"
                    mod_name += name
            klass_path = "%s::%s" % (mod_name, obj.__class__.__name__)
        ARC_Data._dump_string(ARC_Data.__get_class_path(klass_path)) # first the string path because this is required to load the object
        if not ARC_Data.__try_map(ARC_Data._objects, obj): # abort if object has already been mapped
            return
        if hasattr(obj, "_arc_dump"):
            data = obj._arc_dump()
            ARC_Data.__dump_int32(len(data))
            ARC_Data._io.write(data)
        else:
            if hasattr(obj, "_arc_exclude"):
                excludes = obj._arc_exclude
            else:
                excludes = []
            variables = list(set(ARC_Data._request_instance_variables(obj)) - set(excludes))
            ARC_Data.__dump_int32(len(variables))
            for variable in variables:
                ARC_Data._dump_string(variable)
                ARC_Data._dump(getattr(obj, variable))
        
    
    
    @staticmethod
    def _load_none():
        return None
    

    @staticmethod
    def _load_false():
        return False
    

    @staticmethod
    def _load_true():
        return True
    
    
    @staticmethod
    def _load_fixnum():
        return ARC_Data.__load_int32()
    
    
    @staticmethod
    def _load_bignum():
        return ARC_Data.__load_int32() # our C++ implementation uses a "long" of 32 bit
    
    
    @staticmethod
    def _load_float():
        return unpack("<f", ARC_Data._io.read(4))
    
    
    @staticmethod
    def _load_string():
        id_num = ARC_Data.__load_int32()
        if id_num == 0:
            return ""
        obj = ARC_Data.__find_mapped(ARC_Data._strings, id_num)
        if obj != None:
            return obj.clone
        size = ARC_Data.__load_int32()
        obj = ARC_Data._io.read(size)
        ARC_Data.__map(ARC_Data._strings, obj)
        return obj
    
    
    @staticmethod
    def _load_array():
        id_num = ARC_Data.__load_int32()
        if id_num == 0:
            return []
        obj = ARC_Data.__find_mapped(ARC_Data._arrays, id_num)
        if obj != None:
            return obj
        size = ARC_Data.__load_int32()
        obj = []
        ARC_Data.__map(ARC_Data._arrays, obj)
        for i in xrange(size):
            obj.append(ARC_Data._load())
        return obj
    
    
    @staticmethod
    def _load_hash():
        id_num = ARC_Data.__load_int32()
        if id_num == 0:
            return {}
        obj = ARC_Data.__find_mapped(ARC_Data._hashes, id_num)
        if obj != None:
            return obj
        size = ARC_Data.__load_int32()
        obj = {}
        ARC_Data.__map(ARC_Data._hashes, obj)
        for i in xrange(size):
            key = ARC_Data._load() # making sure key is always loaded first
            obj[key] = ARC_Data._load()
        return obj
    
    
    @staticmethod
    def _load_object():
        class_path = ARC_Data._load()
        obj = ARC_Data.__find_mapped(ARC_Data._objects, ARC_Data.__load_int32())
        if obj != None:
            return obj
        classe = ARC_Data.__get_class_object(class_path)
        size = ARC_Data.__load_int32()
        if hasattr(classe, "_arc_load"):
            obj = classe._arc_load(ARC_Data._io.read(size))
            ARC_Data.__map(ARC_Data._objects, obj)
            return obj
        obj = classe.__new__(classe)
        ARC_Data.__map(ARC_Data._objects, obj)
        for i in xrange(size):
            setattr(obj, ARC_Data._load(), ARC_Data._load())
        return obj
    
    

File: engine/external/zer0/src/ARC_Data.cpp
#include <ruby.h>

#include <hltypes/harray.h>
#include <hltypes/hfile.h>
#include <hltypes/hmap.h>
#include <hltypes/hstring.h>

#include "ARC.h"
#include "ARC_Data.h"
#include "ARC_Error.h"
#include "CodeSnippets.h"
#include "zer0.h"

#define __MAP(data, obj) (data += obj)
#define __FIND_MAPPED(data, id) (id < data.size() ? data[id] : Qnil)
#define __DUMP_INT32(obj) (file.dump(obj))
#define __LOAD_INT32 (file.load_int())

namespace zer0
{
	VALUE rb_mARC_Data;

	/****************************************************************************************
	 * Pure C++ code
	 ****************************************************************************************/
	
	hstr ARC_Data::Header;
	hstr ARC_Data::Version;
	hmap<VALUE, unsigned char> ARC_Data::Types;
	hfile ARC_Data::file;
	
	harray<VALUE> ARC_Data::strings;
	harray<VALUE> ARC_Data::arrays;
	harray<VALUE> ARC_Data::hashes;
	harray<VALUE> ARC_Data::objects;
	
	/****************************************************************************************
	 * Ruby Interfacing, Creation, Destruction, Systematics
	 ****************************************************************************************/

	void ARC_Data::init()
	{
		Header = "ARCD";
		Version = "1.0";
		Types[rb_cNilClass] = 0x10;
		Types[rb_cFalseClass] = 0x11;
		Types[rb_cTrueClass] = 0x12;
		Types[rb_cFixnum] = 0x21;
		Types[rb_cBignum] = 0x22;
		Types[rb_cFloat] = 0x23;
		Types[rb_cString] = 0x30;
		Types[rb_cArray] = 0x40;
		Types[rb_cHash] = 0x41;
		Types[rb_cObject] = 0x00;
		ARC_Data::_resetSerializer();
	}

	void ARC_Data::destroy()
	{
	}

	void ARC_Data::createRubyInterface()
	{
		rb_mARC_Data = rb_define_module_under(rb_mARC, "Data");
		rb_define_const(rb_mARC_Data, "Version", rb_str_new2(Version.c_str()));
		rb_define_module_function(rb_mARC_Data, "dump", RUBY_METHOD_FUNC(&ARC_Data::rb_dump), 2);
		rb_define_module_function(rb_mARC_Data, "load", RUBY_METHOD_FUNC(&ARC_Data::rb_load), 1);
	}

	/****************************************************************************************
	 * Utility Load/Dump Methods
	 ****************************************************************************************/

	void ARC_Data::_resetSerializer()
	{
		if (file.is_open())
		{
			file.close();
		}
		strings.clear();
		strings += Qnil;
		arrays.clear();
		arrays += Qnil;
		hashes.clear();
		hashes += Qnil;
		objects.clear();
		objects += Qnil;
	}

	VALUE ARC_Data::__get_class_object(hstr class_path)
	{
		harray<hstr> classes = class_path.split("::");
		VALUE symbol = rb_f_to_sym(rb_str_new2(classes[0].c_str()));
		if (!RTEST(rb_funcall_1(rb_mKernel, "const_defined?", symbol)))
		{
			rb_raise(rb_eARC_Error, ("Class not defined " + classes[0]).c_str());
		}
		classes.pop(0);
		VALUE classe = rb_funcall_1(rb_mKernel, "const_get", symbol);
		foreach (hstr, it, classes)
		{
			symbol = rb_f_to_sym(rb_str_new2((*it).c_str()));
			if (!RTEST(rb_funcall_1(classe, "const_defined?", symbol)))
			{
				rb_raise(rb_eARC_Error, ("Class not defined " + (*it)).c_str());
			}
			classe = rb_funcall_1(classe, "const_get", symbol);
		}
		return classe;
	}

	bool ARC_Data::__try_map(harray<VALUE>& data, VALUE obj)
	{
		int index = data.index_of(obj);
		if (index < 0)
		{
			__DUMP_INT32(data.size());
			data += obj;
			return true;
		}
		__DUMP_INT32(index);
		return false;
	}

	VALUE ARC_Data::__safe_dump(VALUE obj)
	{
		ARC_Data::_dump(obj);
		return Qnil;
	}

	VALUE ARC_Data::__safe_load(VALUE ignored)
	{
		return ARC_Data::_load();
	}

	void ARC_Data::_dump(VALUE obj)
	{
		if (rb_obj_is_instance_of(obj, rb_cNilClass)) return ARC_Data::_dump_nil(obj);
		if (rb_obj_is_instance_of(obj, rb_cFalseClass)) return ARC_Data::_dump_false(obj);
		if (rb_obj_is_instance_of(obj, rb_cTrueClass)) return ARC_Data::_dump_true(obj);
		if (rb_obj_is_instance_of(obj, rb_cFixnum)) return ARC_Data::_dump_fixnum(obj);
		if (rb_obj_is_instance_of(obj, rb_cBignum)) return ARC_Data::_dump_bignum(obj);
		if (rb_obj_is_instance_of(obj, rb_cFloat)) return ARC_Data::_dump_float(obj);
		if (rb_obj_is_instance_of(obj, rb_cString)) return ARC_Data::_dump_string(obj);
		if (rb_obj_is_instance_of(obj, rb_cArray)) return ARC_Data::_dump_array(obj);
		if (rb_obj_is_instance_of(obj, rb_cHash)) return ARC_Data::_dump_hash(obj);
		if (rb_obj_is_instance_of(obj, rb_cObject)) return ARC_Data::_dump_object(obj);
		VALUE class_name = rb_class_name(obj);
		rb_raise(rb_eARC_Error, hsprintf("Error: %s cannot be dumped!", StringValueCStr(class_name)).c_str());
	}
		
	VALUE ARC_Data::_load()
	{
		unsigned char type = file.load_uchar();
		if (type == Types[rb_cNilClass]) return ARC_Data::_load_nil();
		if (type == Types[rb_cFalseClass]) return ARC_Data::_load_false();
		if (type == Types[rb_cTrueClass]) return ARC_Data::_load_true();
		if (type == Types[rb_cFixnum]) return ARC_Data::_load_fixnum();
		if (type == Types[rb_cBignum]) return ARC_Data::_load_bignum();
		if (type == Types[rb_cFloat]) return ARC_Data::_load_float();
		if (type == Types[rb_cString]) return ARC_Data::_load_string();
		if (type == Types[rb_cArray]) return ARC_Data::_load_array();
		if (type == Types[rb_cHash]) return ARC_Data::_load_hash();
		if (type == Types[rb_cObject]) return ARC_Data::_load_object();
		rb_raise(rb_eARC_Error, hsprintf("Error: Unknown type 0x%02X detected!", type).c_str());
		return Qnil;
	}

	void ARC_Data::_dump_nil(VALUE obj)
	{
		file.dump(Types[rb_cNilClass]);
	}

	void ARC_Data::_dump_false(VALUE obj)
	{
		file.dump(Types[rb_cFalseClass]);
	}

	void ARC_Data::_dump_true(VALUE obj)
	{
		file.dump(Types[rb_cTrueClass]);
	}
		
	void ARC_Data::_dump_fixnum(VALUE obj)
	{
		file.dump(Types[rb_cFixnum]);
		__DUMP_INT32(NUM2INT(obj));
	}
		
	void ARC_Data::_dump_bignum(VALUE obj)
	{
		file.dump(Types[rb_cBignum]);
		__DUMP_INT32(NUM2INT(obj)); // the C++ implementation uses a "long" of 32 bit
	}
		
	void ARC_Data::_dump_float(VALUE obj)
	{
		file.dump(Types[rb_cFloat]);
		file.dump((float)NUM2DBL(obj));
	}
		
	void ARC_Data::_dump_string(VALUE obj)
	{
		file.dump(Types[rb_cString]);
		int size = NUM2INT(rb_str_size(obj));
		if (size > 0)
		{
			if (ARC_Data::__try_map(strings, obj))
			{
				file.dump(hstr(StringValueCStr(obj)));
			}
		}
		else
		{
			__DUMP_INT32(0);
		}
	}
		
	void ARC_Data::_dump_array(VALUE obj)
	{
		file.dump(Types[rb_cArray]);
		int size = NUM2INT(rb_ary_size(obj));
		if (size > 0)
		{
			if (ARC_Data::__try_map(arrays, obj))
			{
				__DUMP_INT32(size);
				rb_ary_each_index(obj, i)
				{
					ARC_Data::_dump(rb_ary_entry(obj, i));
				}
			}
		}
		else
		{
			__DUMP_INT32(0);
		}
	}
		
	void ARC_Data::_dump_hash(VALUE obj)
	{
		file.dump(Types[rb_cHash]);
		int size = NUM2INT(rb_hash_size(obj));
		if (size > 0)
		{
			if (ARC_Data::__try_map(hashes, obj))
			{
				__DUMP_INT32(size);
				VALUE keys = rb_funcall_0(obj, "keys");
				VALUE key;
				rb_ary_each_index(keys, i)
				{
					key = rb_ary_entry(keys, i);
					ARC_Data::_dump(key);
					ARC_Data::_dump(rb_hash_aref(obj, key));
				}
			}
		}
		else
		{
			__DUMP_INT32(0);
		}
	}
		
	void ARC_Data::_dump_object(VALUE obj)
	{
		file.dump(Types[rb_cObject]);
		// first the string path because this is required to load the object
		VALUE class_name = rb_class_name(obj);
		file.dump(hstr(StringValueCStr(class_name)));
		if (ARC_Data::__try_map(objects, obj))
		{
			if (RTEST(rb_funcall_1(obj, "respond_to?", rb_str_new2("_arc_dump"))))
			{
				VALUE data = rb_funcall_0(obj, "_arc_dump");
				int size = NUM2INT(rb_str_size(data));
				__DUMP_INT32(size);
				unsigned char* raw_data = (unsigned char*)StringValueCStr(data);
				file.write_raw(raw_data, size);
			}
			else
			{
				VALUE variables = rb_funcall_0(obj, "instance_variables");
				int size = NUM2INT(rb_ary_size(variables));
				__DUMP_INT32(size);
				VALUE variable;
				rb_ary_each_index(variables, i)
				{
					variable = rb_ary_entry(variables, i);
					ARC_Data::_dump_string(rb_funcall_2(rb_f_to_s(variable), "gsub", rb_str_new2("@"), rb_str_new2("")));
					ARC_Data::_dump(rb_funcall_1(obj, "instance_variable_get", variable));
				}
			}
		}
	}

	VALUE ARC_Data::_load_nil()
	{
		return Qnil;
	}

	VALUE ARC_Data::_load_false()
	{
		return Qfalse;
	}

	VALUE ARC_Data::_load_true()
	{
		return Qtrue;
	}
		
	VALUE ARC_Data::_load_fixnum()
	{
		return INT2FIX(__LOAD_INT32);
	}
		
	VALUE ARC_Data::_load_bignum()
	{
		return INT2FIX(__LOAD_INT32); // the C++ implementation uses a "long" of 32 bit
	}
		
	VALUE ARC_Data::_load_float()
	{
		return rb_float_new((double)file.load_float());
	}
	
	VALUE ARC_Data::_load_string()
	{
		int id = __LOAD_INT32;
		if (id == 0)
		{
			return rb_str_new2("");
		}
		VALUE obj = __FIND_MAPPED(strings, id);
		if (!NIL_P(obj))
		{
			return rb_str_new2(StringValueCStr(obj));
		}
		hstr value = file.load_hstr();
		obj = rb_str_new2(value.c_str());
		__MAP(strings, obj);
		return obj;
	}
		
	VALUE ARC_Data::_load_array()
	{
		int id = __LOAD_INT32;
		if (id == 0)
		{
			return rb_ary_new();
		}
		VALUE obj = __FIND_MAPPED(arrays, id);
		if (!NIL_P(obj))
		{
			return obj;
		}
		int size = __LOAD_INT32;
		obj = rb_ary_new();
		__MAP(arrays, obj);
		for_iter (i, 0, size)
		{
			rb_ary_push(obj, ARC_Data::_load());
		}
		return obj;
	}
		
	VALUE ARC_Data::_load_hash()
	{
		int id = __LOAD_INT32;
		if (id == 0)
		{
			return rb_hash_new();
		}
		VALUE obj = __FIND_MAPPED(hashes, id);
		if (!NIL_P(obj))
		{
			return obj;
		}
		int size = __LOAD_INT32;
		obj = rb_hash_new();
		__MAP(hashes, obj);
		VALUE key;
		for_iter (i, 0, size)
		{
			key = ARC_Data::_load(); // making sure key is always loaded first
			rb_hash_aset(obj, key, ARC_Data::_load());
		}
		return obj;
	}
		
	VALUE ARC_Data::_load_object()
	{
		VALUE class_path = ARC_Data::_load();
		int id = __LOAD_INT32;
		VALUE obj = __FIND_MAPPED(objects, id);
		if (!NIL_P(obj))
		{
			return obj;
		}
		VALUE classe = ARC_Data::__get_class_object(hstr(StringValueCStr(class_path)));
		int size = __LOAD_INT32;
		if (rb_funcall_1(classe, "respond_to?", rb_str_new2("_arc_load")))
		{
			unsigned char* data = new unsigned char[size];
			file.read_raw(data, size);
			obj = rb_funcall_1(classe, "_arc_load", rb_str_new((const char*)data, size));
			__MAP(objects, obj);
			delete data;
			return obj;
		}
		obj = rb_funcall_0(classe, "allocate");
		__MAP(objects, obj);
		VALUE variable;
		VALUE symbol;
		VALUE value;
		for_iter (i, 0, size)
		{
			variable = ARC_Data::_load();
			symbol = rb_f_to_sym(rb_str_new2(("@" + hstr(StringValueCStr(variable))).c_str()));
			value = ARC_Data::_load();
			rb_funcall_2(obj, "instance_variable_set", symbol, value);
		}
		return obj;
	}
		
	/****************************************************************************************
	 * Ruby Methods
	 ****************************************************************************************/

	VALUE ARC_Data::rb_dump(VALUE self, VALUE filename, VALUE obj)
	{
		try
		{
			file.open(StringValueCStr(filename));
		}
		catch (hltypes::_file_not_found e)
		{
			RB_RAISE_FILE_NOT_FOUND(StringValueCStr(filename));
		}
		rb_funcall_0(rb_mGC, "disable"); // to prevent GC destroying temp data
		file.dump((unsigned char)Header[0]);
		file.dump((unsigned char)Header[1]);
		file.dump((unsigned char)Header[2]);
		file.dump((unsigned char)Header[3]);
		harray<unsigned char> versions = Version.split(".").cast<int>().cast<unsigned char>();
		file.dump(versions[0]);
		file.dump(versions[1]);
		int exception;
		rb_protect(&ARC_Data::__safe_dump, obj, &exception);
		ARC_Data::_resetSerializer();
		rb_funcall_0(rb_mGC, "enable");
		if (exception != 0)
		{
			rb_jump_tag(exception);
		}
		return Qnil;
	}

	VALUE ARC_Data::rb_load(VALUE self, VALUE filename)
	{
		try
		{
			file.open(StringValueCStr(filename));
		}
		catch (hltypes::_file_not_found e)
		{
			RB_RAISE_FILE_NOT_FOUND(StringValueCStr(filename));
		}
		rb_funcall_0(rb_mGC, "disable"); // to prevent GC destroying temp data
		bool failed = (file.size() < 4);
		char chars[5] = {'\0'};
		hstr header;
		if (!failed)
		{
			chars[0] = (char)file.load_uchar();
			chars[1] = (char)file.load_uchar();
			chars[2] = (char)file.load_uchar();
			chars[3] = (char)file.load_uchar();
			header = hstr(chars);
			failed = (Header != header);
		}
		if (failed)
		{
			ARC_Data::_resetSerializer();
			rb_funcall_0(rb_mGC, "enable");
			rb_raise(rb_eARC_Error, hsprintf("Error: ARC::Data header mismatch! Excepted: \"ARCD\" Found: \"%s\"",
				Header.c_str(), header.c_str()).c_str());
			return Qnil;
		}
		failed = (file.size() < 6);
		unsigned char major;
		unsigned char minor;
		hstr version;
		if (!failed)
		{
			major = file.load_uchar();
			minor = file.load_uchar();
			version = hstr((int)major) + "." + hstr((int)minor);
			failed = (Version != version);
		}
		if (failed)
		{
			ARC_Data::_resetSerializer();
			rb_funcall_0(rb_mGC, "enable");
			rb_raise(rb_eARC_Error, hsprintf("Error: ARC::Data version mismatch! Excepted: \"%s\" Found: \"%s\"",
				Version.c_str(), version.c_str()).c_str());
			return Qnil;
		}
		int exception;
		VALUE data = rb_protect(&ARC_Data::__safe_load, Qnil, &exception);
		ARC_Data::_resetSerializer();
		rb_funcall_0(rb_mGC, "enable");
		if (exception != 0)
		{
			rb_jump_tag(exception);
			return Qnil;
		}
		return data;
	}

}
